In [1]:
import numpy as np
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense

In [5]:
num_words = 10000
maxlen = 200

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=num_words)
X_train = pad_sequences(X_train, maxlen=maxlen)
X_test = pad_sequences(X_test, maxlen=maxlen)

RNN

In [6]:
model_rnn = Sequential()
model_rnn.add(Embedding(num_words, 32))
model_rnn.add(SimpleRNN(64))
model_rnn.add(Dense(1, activation='sigmoid'))

model_rnn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_rnn.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_2 (SimpleRNN)        │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [7]:
model_rnn.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_rnn.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 7s 31ms/step - accuracy: 0.5700 - loss: 0.6655 - val_accuracy: 0.8188 - val_loss: 0.4177
Epoch 2/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.7911 - loss: 0.4494 - val_accuracy: 0.7981 - val_loss: 0.4534
Epoch 3/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.8500 - loss: 0.3555 - val_accuracy: 0.8369 - val_loss: 0.3853
Epoch 4/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.9028 - loss: 0.2502 - val_accuracy: 0.8090 - val_loss: 0.4226
Epoch 5/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.9328 - loss: 0.1823 - val_accuracy: 0.8086 - val_loss: 0.4655
Epoch 6/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.9676 - loss: 0.0997 - val_accuracy: 0.8289 - val_loss: 0.4785
Epoch 7/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.9872 - loss: 0.0517 - val_accuracy: 0.8197 - val_loss: 0.5991
Epoch 8/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 6s 31ms/step - accuracy: 0.9886 - loss: 0.0397 - val_accu

RNN

10 епох, виконання 1 хвилина, точність 79.06%. Під час тренування спостерігався overfitting. Пік точності було досягнуто на 9 епосі - 99,85%

LSTM

In [8]:
model_lstm = Sequential()
model_lstm.add(Embedding(num_words, 32))
model_lstm.add(LSTM(64))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_lstm.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_3 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm (LSTM)                     │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [9]:
model_lstm.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_lstm.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 20s 94ms/step - accuracy: 0.6735 - loss: 0.6077 - val_accuracy: 0.8454 - val_loss: 0.3581
Epoch 2/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 92ms/step - accuracy: 0.8933 - loss: 0.2709 - val_accuracy: 0.8407 - val_loss: 0.3625
Epoch 3/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 91ms/step - accuracy: 0.9209 - loss: 0.2110 - val_accuracy: 0.8722 - val_loss: 0.3151
Epoch 4/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 91ms/step - accuracy: 0.9454 - loss: 0.1541 - val_accuracy: 0.8654 - val_loss: 0.3741
Epoch 5/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 91ms/step - accuracy: 0.9626 - loss: 0.1150 - val_accuracy: 0.8390 - val_loss: 0.3896
Epoch 6/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 91ms/step - accuracy: 0.9632 - loss: 0.1088 - val_accuracy: 0.8621 - val_loss: 0.4145
Epoch 7/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 91ms/step - accuracy: 0.9743 - loss: 0.0793 - val_accuracy: 0.8476 - val_loss: 0.4790
Epoch 8/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 18s 90ms/step - accuracy: 0.9833 - loss: 0.0572 - 

LSTM

10 епох, 3 хвилини, точність 84.94%. Спостерігається незначний спад точності після початкового підвищення

Двостороння мережа (Bidirectional)

In [10]:
model_bidirectional = Sequential()
model_bidirectional.add(Embedding(num_words, 32))
model_bidirectional.add(Bidirectional(LSTM(64)))
model_bidirectional.add(Dense(1, activation='sigmoid'))

model_bidirectional.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_bidirectional.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ bidirectional (Bidirectional)   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [11]:
model_bidirectional.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_bidirectional.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 29s 137ms/step - accuracy: 0.6481 - loss: 0.5973 - val_accuracy: 0.8644 - val_loss: 0.3245
Epoch 2/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 133ms/step - accuracy: 0.8984 - loss: 0.2647 - val_accuracy: 0.8731 - val_loss: 0.3133
Epoch 3/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 132ms/step - accuracy: 0.9348 - loss: 0.1819 - val_accuracy: 0.8704 - val_loss: 0.3219
Epoch 4/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 133ms/step - accuracy: 0.9472 - loss: 0.1502 - val_accuracy: 0.8660 - val_loss: 0.3416
Epoch 5/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 133ms/step - accuracy: 0.9580 - loss: 0.1206 - val_accuracy: 0.8646 - val_loss: 0.3782
Epoch 6/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 133ms/step - accuracy: 0.9660 - loss: 0.1021 - val_accuracy: 0.8519 - val_loss: 0.4737
Epoch 7/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 134ms/step - accuracy: 0.9672 - loss: 0.0989 - val_accuracy: 0.8599 - val_loss: 0.4185
Epoch 8/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 26s 135ms/step - accuracy: 0.9750 - loss: 0

Bidirectional

10 епох, виконання 5 хвилин, точність 85.08%. Попри зниження точності після перших кількох епох, модель Bidirectional все ще показує достатньо високу точність

Глибока мережа (Deep)

In [12]:
model_deep = Sequential()
model_deep.add(Embedding(num_words, 32))
model_deep.add(LSTM(64, return_sequences=True))
model_deep.add(LSTM(64))
model_deep.add(Dense(1, activation='sigmoid'))

model_deep.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_deep.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_5 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_2 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_3 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [13]:
model_deep.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_deep.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 45s 217ms/step - accuracy: 0.6834 - loss: 0.5463 - val_accuracy: 0.8747 - val_loss: 0.3089
Epoch 2/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 46s 237ms/step - accuracy: 0.9078 - loss: 0.2415 - val_accuracy: 0.8673 - val_loss: 0.3104
Epoch 3/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 44s 225ms/step - accuracy: 0.9363 - loss: 0.1751 - val_accuracy: 0.8651 - val_loss: 0.3786
Epoch 4/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 43s 218ms/step - accuracy: 0.9559 - loss: 0.1248 - val_accuracy: 0.8628 - val_loss: 0.3537
Epoch 5/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 41s 210ms/step - accuracy: 0.9649 - loss: 0.1022 - val_accuracy: 0.8484 - val_loss: 0.4676
Epoch 6/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 41s 211ms/step - accuracy: 0.9754 - loss: 0.0757 - val_accuracy: 0.8501 - val_loss: 0.4390
Epoch 7/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 41s 211ms/step - accuracy: 0.9807 - loss: 0.0591 - val_accuracy: 0.8563 - val_loss: 0.5260
Epoch 8/10
196/196 ━━━━━━━━━━━━━━━━━━━━ 43s 218ms/step - accuracy: 0.9877 - loss: 0

Deep

10 епох, виконання 8 хвилин, точність 84.26%. Після певного зростання точності протягом перших епох, точність на валідаційному наборі стає стійкою та показує добрі результати


Висновки

Після тренування моделей протягом 5 епох

LSTM та двостороння мережа Bidirectional показують кращі результати, оскільки вони можуть краще усвідомлювати довгострокові залежності в тексті. Однак, різниця у точності між цими моделями є несуттєвою.

Загальні

RNN: При збільшенні кількості епох з 5 до 10 спостерігається overfitting моделі, що проявляється у зниженні точності на валідаційному наборі даних. При 7 епохах досягається пік точності, але загальна точність (77.06%) є нижчою в порівнянні з іншими моделями.

LSTM: При 5 епохах модель досягає високої точності (85.98%), але при подовженні тренування до 10 епох точність незначно знижується (84.84%). Це може свідчити про наявність незначного overfitting після певної кількості епох.

Bidirectional: Модель показує добрі результати при обох кількостях епох (85.76% при 5 епохах і 84.94% при 10 епохах). Хоча точність знижується після перших кількох епох, вона все ще залишається на високому рівні, що свідчить про ефективність моделі.

Deep: Точність моделі Deep також залишається на високому рівні незалежно від кількості епох (85.55% при 5 епохах і 84.61% при 10 епохах). Після певного зростання точності протягом перших епох, модель стає стійкою і показує добрі результати.

Загалом, можна зробити висновок, що моделі LSTM, Bidirectional і Deep показують добрі результати у класифікації рецензій, з високою точністю навіть після певного зниження в деяких випадках. Проте, Bidirectional і особливо Deep вимагають значно більше часу для навчання
